source/_posts/2017-06-30-uefi-slash-edk-ii-corebootpkg-on-pc-engines-apu2.markdown

```diff
@@ -0,0 +1,335 @@
+
+It is important to deselect secondary payloads like `memtest86+` and
+`sortbootorder` to avoid compilation issues.
+
+### EDK2 developer's script
+
+For all interested in EDK2 code development I strongly advise to follow [Laszlo's  guide](https://github.com/tianocore/tianocore.github.io/wiki/Laszlo's-unkempt-git-guide-for-edk2-contributors-and-maintainers).
+
+In case of APU2 useful script may be:
+
+```sh
+#!/bin/bash
+build -a IA32 -p CorebootPayloadPkg/CorebootPayloadPkgIa32.dsc -b DEBUG -t GCC5
+cp Build/CorebootPayloadPkgIA32/DEBUG_GCC5/FV/UEFIPAYLOAD.fd ../apu2_fw_rel/apu2/coreboot/
+cd ../apu2_fw_rel
+./apu2/apu2-documentation/scripts/apu2_fw_rel.sh build-ml
+read -n 1 -s -p "Press any key to continue with flashing recent build ..."
+./apu2/apu2-documentation/scripts/apu2_fw_rel.sh flash-ml root@192.168.0.105
+cd ../edk2
+```
+
+I added prompt for flashing since it happens to forget remove recovery flash,
+what may lead to blocking further development since both SPIs will contain not
+bootable firmware.
+
+
+## Hacking UEFI payload to boot on APU2
+
+### CbSupportDxe assert
+
+First problem I faced was assert in `CbSupportDxe.c` related to adding 1MB
+memory for LAPIC. Reservation happens in CbSupportDxe entry point.
+
+```
+Loading driver at 0x000CFC25000 EntryPoint=0x000CFC2595B CbSupportDxe.efi
+InstallProtocolInterface: BC62157E-3E33-4FEC-9920-2D3B36D750DF CFC42190
+ProtectUefiImageCommon - 0xCFC42928
+  - 0x00000000CFC25000 - 0x0000000000002580
+PROGRESS CODE: V03040002 I0
+Failed to add memory space :0xFEE00000 0x100000
+
+ASSERT_EFI_ERROR (Status = Access Denied)
+
+DXE_ASSERT!: [CbSupportDxe] /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/edk2/CorebootModulePkg/CbSupportDxe/CbSupportDxe.c (57): !EFI_ERROR (Status)
+```
+
+Interestingly CbSupportDxe seems to not have problem with finding GUID HOB,
+which is right after asserting reservation. It also passes installation of ACPI
+and SMBIOS table. I just commented that reservation and moved forward to see
+what will happen.
+
+### PcRtcEntry assert
+
+Next problem happens during initialization of RTC:
+
+```
+Loading driver at 0x000CFE50000 EntryPoint=0x000CFE507DB PcRtc.efi
+InstallProtocolInterface: BC62157E-3E33-4FEC-9920-2D3B36D750DF CFC2AE90
+ProtectUefiImageCommon - 0xCFC2AB28
+  - 0x00000000CFE50000 - 0x0000000000003980
+  PROGRESS CODE: V03040002 I0
+  ASSERT_EFI_ERROR (Status = Device Error)
+
+  DXE_ASSERT!: [PcRtc] /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/edk2/PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcRtcEntry.c (143): !EFI_ERROR (Status)
+```
+
+Unfortunately Real Time Clock is required architecture protocol and cannot be
+omitted.
+
+First problem with this code was an incorrect state of `Valid RAM and Time` (`VRT`) bit in
+RTC Date Alarm register (aka Register D).
+
+By checking BKDG I was not able to find issue with RTC. Reading all registers was
```

Comment:
Would be nice to add explication of the BKDG acronym, it could be not known to people unfamiliar with BIOS development.